conjure-core/src/main/java/com/palantir/parsec/Parsers.java

```diff
@@ -108,28 +108,59 @@ public T parse(ParserState input) throws ParseException {
     /**
      * Parse either firstOption or iterate through otherOptions and return result.
      *
+     * Exceptions are caught while trying options - if no options succeed, then the last exception thrown is re-thrown
+     * as a {@code ParseError}.
+     *
      * @param <T> the type the parsers return
      * @param firstOption the first parser to try
      * @param otherOptions additional parsers to try
      * @return the specified parser
      */
     @SafeVarargs
     public static <T> Parser<T> or(final Parser<? extends T> firstOption, final Parser<? extends T>... otherOptions) {
-        return new Parser<T>() {
-            @Override
-            public T parse(ParserState input) throws ParseException {
-                T result = gingerly(firstOption).parse(input);
+        return input -> {
+            T result;
+            Exception exception = null;
+
+            input.mark();
+            try {
+                result = gingerly(firstOption).parse(input);
+            } catch (RuntimeException e) {
+                exception = e;
+                input.rewind();
+                result = null;
+            }
+
+            if (result != null) {
+                return result;
+            }
+            for (Parser<? extends T> nextOption : otherOptions) {
+                input.mark();
+                try {
+                    result = gingerly(nextOption).parse(input);
+                } catch (RuntimeException e) {
+                    if (exception == null) {
+                        exception = e;
+                    } else {
+                        exception.addSuppressed(e);
+                    }
+                    input.rewind();
+                }
+
                 if (result != null) {
                     return result;
                 }
-                for (Parser<? extends T> nextOption : otherOptions) {
-                    result = gingerly(nextOption).parse(input);
-                    if (result != null) {
-                        return result;
-                    }
+            }
+
+            if (exception != null) {
+                ParseException toThrow = new ParseException(exception.getMessage(), input);
```

Comment:
Let's add a `ParseException` constructor overload so we don't have to copy over the suppressed values, they're included in the cause. This way we don't lose the stack trace data from the initial exception.